File: trx_failures.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

// dotnet run --file trx_failures.cs -- results.trx -o failures.md

using System.Text;
using System.Text.RegularExpressions;
using System.Xml.Linq;

return TrxFailures.Run(args);

static partial class TrxFailures
{
    private static readonly XNamespace s_trxNamespace = "http://microsoft.com/schemas/VisualStudio/TeamTest/2010";

    public static int Run(string[] args)
    {
        if (!TryParseArguments(args, out var trxPaths, out var outputPath))
        {
            PrintUsage();
            return 1;
        }

        var failures = new List<TestFailure>();
        foreach (var path in trxPaths)
        {
            failures.AddRange(Collect(path));
        }

        var uniqueFailures = failures
            .OrderBy(failure => failure.Name, StringComparer.Ordinal)
            .ThenBy(failure => failure.File, StringComparer.Ordinal)
            .ThenBy(failure => failure.Line ?? 0)
            .DistinctBy(failure => (failure.Name, failure.File, failure.Line));

        var markdown = FormatMarkdown(uniqueFailures);
        if (outputPath is not null)
        {
            File.WriteAllText(outputPath, markdown + Environment.NewLine, new UTF8Encoding(encoderShouldEmitUTF8Identifier: false));
        }
        else
        {
            Console.WriteLine(markdown);
        }

        return 0;
    }

    private static bool TryParseArguments(string[] args, out List<string> trxPaths, out string? outputPath)
    {
        trxPaths = [];
        outputPath = null;
        var parseOptions = true;

        for (var i = 0; i < args.Length; i++)
        {
            var argument = args[i];
            if (parseOptions && argument == "--")
            {
                parseOptions = false;
            }
            else if (parseOptions && argument is "-o" or "--output")
            {
                if (++i >= args.Length)
                {
                    return false;
                }

                outputPath = args[i];
            }
            else if (parseOptions && argument.StartsWith("--output=", StringComparison.Ordinal))
            {
                outputPath = argument["--output=".Length..];
                if (outputPath.Length == 0)
                {
                    return false;
                }
            }
            else if (parseOptions && argument.StartsWith("-o", StringComparison.Ordinal) && argument.Length > 2)
            {
                outputPath = argument[2..];
            }
            else if (parseOptions && argument.StartsWith('-', StringComparison.Ordinal))
            {
                return false;
            }
            else
            {
                trxPaths.Add(argument);
            }
        }

        return trxPaths.Count > 0;
    }

    private static IEnumerable<TestFailure> Collect(string trxPath)
    {
        var root = XDocument.Load(trxPath).Root
            ?? throw new InvalidDataException($"TRX file '{trxPath}' does not have a root element.");

        var definitions = new Dictionary<string, TestDefinition>(StringComparer.Ordinal);
        foreach (var unitTest in FindAll(root, "UnitTest"))
        {
            var id = (string?)unitTest.Attribute("id");
            if (id is null)
            {
                continue;
            }

            var method = Find(unitTest, "TestMethod");
            definitions[id] = new TestDefinition(
                (string?)method?.Attribute("className"),
                (string?)unitTest.Attribute("storage") ?? (string?)method?.Attribute("codeBase"));
        }

        foreach (var result in FindAll(root, "UnitTestResult"))
        {
            if (!string.Equals((string?)result.Attribute("outcome"), "failed", StringComparison.OrdinalIgnoreCase))
            {
                continue;
            }

            var testId = (string?)result.Attribute("testId");
            definitions.TryGetValue(testId ?? "", out var definition);

            var output = Find(result, "Output");
            var errorInfo = output is null ? null : Find(output, "ErrorInfo");
            var stack = GetText(errorInfo, "StackTrace");
            var message = string.Join(' ', GetText(errorInfo, "Message").Split((string[]?)null, StringSplitOptions.RemoveEmptyEntries));
            var (file, line) = LocationFromStack(stack);

            yield return new TestFailure(
                (string?)result.Attribute("testName") ?? "",
                definition?.ClassName,
                definition?.Storage,
                file,
                line,
                message);
        }
    }

    private static string FormatMarkdown(IEnumerable<TestFailure> failures)
    {
        var lines = new List<string>
        {
            "| Test | Location |",
            "| --- | --- |",
        };

        foreach (var failure in failures)
        {
            var location = failure.File is not null
                ? $"{failure.File}:{failure.Line}"
                : failure.ClassName ?? failure.Storage ?? "unknown location";
            lines.Add($"| `{EscapeTableCell(failure.Name)}` | {EscapeTableCell(location)} |");
        }

        return string.Join(Environment.NewLine, lines);
    }

    private static string EscapeTableCell(string value)
    {
        return value.Replace("|", "\\|", StringComparison.Ordinal);
    }

    private static (string? File, int? Line) LocationFromStack(string stack)
    {
        var match = StackLocationRegex().Match(stack);
        return match.Success
            ? (match.Groups["file"].Value.Trim(), int.Parse(match.Groups["line"].Value))
            : (null, null);
    }

    private static IReadOnlyList<XElement> FindAll(XElement root, string tag)
    {
        var namespaced = root.Descendants(s_trxNamespace + tag).ToList();
        return namespaced.Count > 0 ? namespaced : root.Descendants(tag).ToList();
    }

    private static XElement? Find(XElement element, string tag)
    {
        return element.Element(s_trxNamespace + tag) ?? element.Element(tag);
    }

    private static string GetText(XElement? element, string tag)
    {
        return element is null ? "" : Find(element, tag)?.FirstNode is XText text ? text.Value : "";
    }

    private static void PrintUsage()
    {
        Console.Error.WriteLine("Usage: dotnet run --file trx_failures.cs -- <file.trx> [more.trx ...] [-o OUT.md]");
    }

    [GeneratedRegex(@"\sin\s(?<file>.+?):line\s(?<line>\d+)", RegexOptions.IgnoreCase)]
    private static partial Regex StackLocationRegex();

    private sealed record TestDefinition(string? ClassName, string? Storage);

    private sealed record TestFailure(
        string Name,
        string? ClassName,
        string? Storage,
        string? File,
        int? Line,
        string Message);
}

File: trx_failures.py
#!/usr/bin/env python3
"""Extract failing tests from a VSTest .trx file as a Markdown list.

Usage:
    trx_failures.py <file.trx> [more.trx ...] [-o OUT.md]

Output format (one bullet per failing test):
    - `Fully.Qualified.Test.Name` — path/to/File.cs:123
"""

import argparse
import re
import sys
import xml.etree.ElementTree as ET

NS = {"t": "http://microsoft.com/schemas/VisualStudio/TeamTest/2010"}

# " at Ns.Type.Method() in C:\src\File.cs:line 42"  (also matches localized-free ":line N")
STACK_LOCATION = re.compile(r"\sin\s(?P<file>.+?):line\s(?P<line>\d+)", re.IGNORECASE)


def _findall(root, tag):
    """Find elements by tag with or without the TRX namespace."""
    found = root.findall(f".//t:{tag}", NS)
    return found if found else root.findall(f".//{tag}")


def _find(elem, tag):
    found = elem.find(f"t:{tag}", NS)
    return found if found is not None else elem.find(tag)


def _text(elem, tag):
    child = _find(elem, tag) if elem is not None else None
    return child.text if child is not None and child.text else ""


def location_from_stack(stack):
    """Return (file, line) of the deepest test-owned frame, or (None, None)."""
    matches = list(STACK_LOCATION.finditer(stack or ""))
    if not matches:
        return None, None
    m = matches[0]
    return m.group("file").strip(), int(m.group("line"))


def collect(trx_path):
    root = ET.parse(trx_path).getroot()

    # testId -> (className, codeBase) from the test definitions
    definitions = {}
    for unit_test in _findall(root, "UnitTest"):
        method = _find(unit_test, "TestMethod")
        definitions[unit_test.get("id")] = (
            method.get("className") if method is not None else None,
            unit_test.get("storage") or (method.get("codeBase") if method is not None else None),
        )

    failures = []
    for result in _findall(root, "UnitTestResult"):
        if (result.get("outcome") or "").lower() != "failed":
            continue

        test_id = result.get("testId")
        class_name, storage = definitions.get(test_id, (None, None))
        name = result.get("testName") or ""

        output = _find(result, "Output")
        error_info = _find(output, "ErrorInfo") if output is not None else None
        stack = _text(error_info, "StackTrace")
        message = " ".join((_text(error_info, "Message") or "").split())

        file_path, line = location_from_stack(stack)
        failures.append(
            {
                "name": name,
                "class": class_name,
                "storage": storage,
                "file": file_path,
                "line": line,
                "message": message,
            }
        )
    return failures


def format_markdown(failures):
    lines = []
    for f in failures:
        if f["file"]:
            location = f"{f['file']}:{f['line']}"
        elif f["class"]:
            location = f["class"]
        elif f["storage"]:
            location = f["storage"]
        else:
            location = "unknown location"
        lines.append(f"- `{f['name']}` - {location}")
    return "\n".join(lines)


def main(argv=None):
    parser = argparse.ArgumentParser(description="Extract failing tests from .trx files as Markdown.")
    parser.add_argument("trx", nargs="+", help="Path(s) to .trx file(s)")
    parser.add_argument("-o", "--output", help="Write Markdown here instead of stdout")
    args = parser.parse_args(argv)

    failures = []
    for path in args.trx:
        failures.extend(collect(path))

    # Stable order, de-duplicated by (name, file, line).
    seen = set()
    unique = []
    for f in sorted(failures, key=lambda x: (x["name"] or "", x["file"] or "", x["line"] or 0)):
        key = (f["name"], f["file"], f["line"])
        if key not in seen:
            seen.add(key)
            unique.append(f)

    markdown = format_markdown(unique)
    if args.output:
        with open(args.output, "w", encoding="utf-8") as handle:
            handle.write(markdown + "\n")
    else:
        print(markdown)

    return 0


if __name__ == "__main__":
    sys.exit(main())